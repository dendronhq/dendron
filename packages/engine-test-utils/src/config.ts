import { DConfig, readYAML, writeYAML } from "@dendronhq/common-server";
import { IntermediateDendronConfig } from "@dendronhq/common-all";

export class TestConfigUtils {
  static getConfig = (opts: { wsRoot: string }) => {
    const configPath = DConfig.configPath(opts.wsRoot);
    const config = readYAML(configPath) as IntermediateDendronConfig;
    return config;
  };

  static withConfig = (
    func: (config: IntermediateDendronConfig) => IntermediateDendronConfig,
    opts: { wsRoot: string }
  ) => {
    const config = TestConfigUtils.getConfig(opts);

    const newConfig = func(config);
    TestConfigUtils.writeConfig({ config: newConfig, wsRoot: opts.wsRoot });
    return newConfig;
  };

  static writeConfig = (opts: {
    config: IntermediateDendronConfig;
    wsRoot: string;
  }) => {
    const configPath = DConfig.configPath(opts.wsRoot);
    return writeYAML(configPath, opts.config);
  };
}
