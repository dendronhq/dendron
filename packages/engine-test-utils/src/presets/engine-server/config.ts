import { CONSTANTS, ConfigUtils } from "@dendronhq/common-all";
import { DConfig, readYAML, writeYAML } from "@dendronhq/common-server";
import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";
import path from "path";

function genDefaultConfig() {
  return ConfigUtils.genDefaultConfig();
}

const WRITE = {
  NEW_CONFIG: new TestPresetEntryV4(async ({ wsRoot }) => {
    const config = genDefaultConfig();
    ConfigUtils.setPublishProp(config, "copyAssets", false);
    await DConfig.writeConfig({ wsRoot, config });
    const cpath = path.join(wsRoot, CONSTANTS.DENDRON_CONFIG_FILE);
    const configOnFile = readYAML(cpath);
    return [
      {
        actual: configOnFile,
        expected: config,
      },
    ];
  }),
};

const GET = {
  DEFAULT_CONFIG: new TestPresetEntryV4(async ({ wsRoot }) => {
    const cpath = path.join(wsRoot, CONSTANTS.DENDRON_CONFIG_FILE);
    writeYAML(cpath, genDefaultConfig());
    const config = DConfig.readConfigSync(wsRoot);
    return [
      {
        actual: config,
        expected: genDefaultConfig(),
      },
    ];
  }),
};

export const ENGINE_CONFIG_PRESETS = {
  WRITE,
  GET,
};
