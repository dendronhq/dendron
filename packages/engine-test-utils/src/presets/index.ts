import { DVault, VaultUtils } from "@dendronhq/common-all";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { WorkspaceOpts } from "../engine";

export * from "./engine-server";
export * from "./pods-core";
export * from "./vaults";

export enum SETUP_HOOK_KEYS {
  /**
   * alpha: link(beta)
   * beta: link(alpha)
   */
  WITH_LINKS = "WITH_LINKS",
}

const createLink = (fname: string, opts: { vaultPrefix?: DVault }) => {
  let cVaultPrefix = "";
  if (opts.vaultPrefix) {
    cVaultPrefix = VaultUtils.toURIPrefix(opts.vaultPrefix) + "/";
  }
  return `[[${cVaultPrefix}${fname}]]`;
};

export async function callSetupHook(
  key: SETUP_HOOK_KEYS,
  opts: {
    workspaceType: "single" | "multi";
    withVaultPrefix?: boolean;
  } & WorkspaceOpts
) {
  const { workspaceType, vaults, wsRoot, withVaultPrefix } = opts;
  const isMultiVault = workspaceType !== "single";
  const cVaults = isMultiVault ? vaults : [vaults[0], vaults[0]];

  // WITH LINKS
  if (key === SETUP_HOOK_KEYS.WITH_LINKS) {
    const link1 = createLink("beta", {
      vaultPrefix: withVaultPrefix ? cVaults[1] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "alpha",
      wsRoot,
      vault: cVaults[0],
      body: link1,
    });
    const link2 = createLink("alpha", {
      vaultPrefix: withVaultPrefix ? cVaults[0] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "beta",
      wsRoot,
      vault: cVaults[1],
      body: link2,
    });
    if (isMultiVault) {
      await NoteTestUtilsV4.createNote({
        fname: "beta",
        wsRoot,
        vault: cVaults[0],
        body: link2,
        genRandomId: true,
      });
    }
  } else {
    throw Error("not supported key");
  }
}
