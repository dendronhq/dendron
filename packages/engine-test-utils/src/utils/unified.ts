import { checkString } from ".";

export class TestUnifiedUtils {
  static verifyPrivateLink = ({
    contents,
    value,
  }: {
    contents: string;
    value: string;
  }) => {
    return checkString(contents, "color: brown", value + " (Private)");
  };
}
