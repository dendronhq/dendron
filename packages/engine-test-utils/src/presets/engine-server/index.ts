import { ENGINE_CONFIG_PRESETS } from "./config";
import { ENGINE_DELETE_PRESETS } from "./delete";
import { ENGINE_INFO_PRESETS } from "./info";
import { ENGINE_INIT_PRESETS } from "./init";
import { ENGINE_GET_NOTE_BLOCKS_PRESETS } from "./getNoteBlocks";
import NOTE_REF from "./note-refs";
import { ENGINE_QUERY_PRESETS } from "./query";
import { ENGINE_RENAME_PRESETS } from "./rename";
import { ENGINE_BULK_WRITE_NOTES_PRESETS } from "./bulkWriteNotes";
import { ENGINE_RENDER_PRESETS } from "./render";
import { ENGINE_WRITE_PRESETS, ENGINE_WRITE_PRESETS_MULTI } from "./write";
import _ from "lodash";
import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";

export { ENGINE_HOOKS, ENGINE_HOOKS_BASE, ENGINE_HOOKS_MULTI } from "./utils";
export { ENGINE_RENAME_PRESETS };
export { ENGINE_QUERY_PRESETS };
export { ENGINE_WRITE_PRESETS };
export { ENGINE_CONFIG_PRESETS };

export const ENGINE_SERVER = {
  NOTE_REF,
  ENGINE_WRITE_PRESETS,
  ENGINE_INIT_PRESETS,
  ENGINE_DELETE_PRESETS,
  ENGINE_INFO_PRESETS,
  ENGINE_RENAME_PRESETS,
  ENGINE_GET_NOTE_BLOCKS_PRESETS,
  ENGINE_QUERY_PRESETS,
  ENGINE_BULK_WRITE_NOTES_PRESETS,
  ENGINE_RENDER_PRESETS,
};

type TestPresetEntry = TestPresetEntryV4;
type TestPresetDict = { [key: string]: TestPresetEntry };

/**
 * 
 @example
 *  test("", async () => {
 *    const TestCase= getPreset({presets: ENGINE_PRESETS, nodeType, presetName: "init", key: 'BAD_SCHEMA'})
 *    const { testFunc, ...opts } = TestCase;;
 *    await runEngineTestV5(testFunc, { ...opts, expect });
 *});
 * @param param0 
 * @returns 
 */
export const getPreset = ({
  presets,
  presetName,
  nodeType,
  key,
}: {
  presets: typeof ENGINE_PRESETS;
  presetName: string;
  nodeType: "SCHEMAS" | "NOTES";
  key: string;
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  const out = _.get(ent.presets[nodeType], key);
  if (!out) {
    throw Error(`no key ${key} found in ${presetName}`);
  }
  return out;
};

export const getPresetMulti = ({
  presets,
  presetName,
  nodeType,
  key,
}: {
  presets: typeof ENGINE_PRESETS_MULTI;
  presetName: string;
  nodeType: "NOTES";
  key: string;
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  const out = _.get(ent.presets[nodeType], key);
  if (!out) {
    throw Error(`no key ${key} found in ${presetName}`);
  }
  return out;
};

export const getPresetGroup = ({
  presets,
  presetName,
  nodeType,
}: {
  presets: typeof ENGINE_PRESETS;
  presetName: string;
  nodeType: "SCHEMAS" | "NOTES";
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  return ent.presets[nodeType] as TestPresetDict;
};

// ^iygzn9r2758w
export const ENGINE_PRESETS = [
  { name: "write", presets: ENGINE_SERVER.ENGINE_WRITE_PRESETS },
  {
    name: "bulkWriteNotes",
    presets: ENGINE_SERVER.ENGINE_BULK_WRITE_NOTES_PRESETS,
  },
  { name: "delete", presets: ENGINE_SERVER.ENGINE_DELETE_PRESETS },
  { name: "rename", presets: ENGINE_SERVER.ENGINE_RENAME_PRESETS },
  { name: "init", presets: ENGINE_SERVER.ENGINE_INIT_PRESETS },
  { name: "query", presets: ENGINE_SERVER.ENGINE_QUERY_PRESETS },
  { name: "info", presets: ENGINE_SERVER.ENGINE_INFO_PRESETS },
  {
    name: "getNoteBlocks",
    presets: ENGINE_SERVER.ENGINE_GET_NOTE_BLOCKS_PRESETS,
  },
  { name: "render", presets: ENGINE_SERVER.ENGINE_RENDER_PRESETS },
];

export const ENGINE_PRESETS_MULTI = [
  { name: "write", presets: ENGINE_WRITE_PRESETS_MULTI },
  { name: "delete", presets: ENGINE_SERVER.ENGINE_DELETE_PRESETS },
  { name: "rename", presets: ENGINE_SERVER.ENGINE_RENAME_PRESETS },
  { name: "init", presets: ENGINE_SERVER.ENGINE_INIT_PRESETS },
  { name: "query", presets: ENGINE_SERVER.ENGINE_QUERY_PRESETS },
];
