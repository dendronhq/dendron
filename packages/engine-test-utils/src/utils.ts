import {
  CONSTANTS,
  DendronConfig,
  VaultUtils,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings,
} from "@dendronhq/common-all";
import { readYAML } from "@dendronhq/common-server";
import { AssertUtils } from "@dendronhq/common-test-utils";
import { DConfig, Git } from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";

export async function checkString(body: string, ...match: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      match,
    })
  ).toBeTruthy();
}

export async function checkNotInString(body: string, ...nomatch: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      nomatch,
    })
  ).toBeTruthy();
}

const getWorkspaceFolders = (wsRoot: string) => {
  const wsPath = path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME);
  const settings = fs.readJSONSync(wsPath) as WorkspaceSettings;
  return _.toArray(settings.folders);
};

export function checkVaults(opts: WorkspaceOpts, expect: any) {
  const { wsRoot, vaults } = opts;
  const configPath = DConfig.configPath(opts.wsRoot);
  const config = readYAML(configPath) as DendronConfig;
  expect(_.sortBy(config.vaults, ["fsPath", "workspace"])).toEqual(
    _.sortBy(vaults, ["fsPath", "workspace"])
  );
  const wsFolders = getWorkspaceFolders(wsRoot);
  expect(wsFolders).toEqual(
    vaults.map((ent) => {
      const out: WorkspaceFolderRaw = { path: VaultUtils.getRelPath(ent) };
      if (ent.name) {
        out.name = ent.name;
      }
      return out;
    })
  );
}

export class GitTestUtils {
  static async createRepoForWorkspace(wsRoot: string) {
    const git = new Git({ localUrl: wsRoot });
    await git.init();
    await git.add("dendron.yml");
    await git.commit({ msg: "init" });
  }

  /** Creates a "bare" git repository, to be used as the remote for a workspace.
   *
   * You'll probably want to just use `createRepoForRemoteWorkspace` instead, but this is provided if you are testing something it can't handle.
   */
  static async remoteCreate(remoteDir: string) {
    const git = new Git({ localUrl: remoteDir, bare: true });
    await git.init();
  }

  /** Adds a bare repository created with `createRemote` as the remote for the workspace.
   *
   * You'll probably want to just use `createRepoForRemoteWorkspace` instead, but this is provided if you are testing something it can't handle.
   */
  static async remoteAdd(wsRoot: string, remoteDir: string) {
    const git = new Git({ localUrl: wsRoot, remoteUrl: remoteDir });
    await git.remoteAdd();
    // Need to push to be able to set up remote tracking branch
    await git.push({ remote: "origin", branch: await git.getCurrentBranch() });
  }

  /** Set up a workspace with a remote, intended to be used when testing pull or push functionality.
   *
   * @param wsRoot Directory where the workspace will be stored.
   * @param remoteDir Directory where the remote will be stored. The workspace will pull and push to this remote.
   */
  static async createRepoForRemoteWorkspace(wsRoot: string, remoteDir: string) {
    await this.createRepoForWorkspace(wsRoot);
    await this.remoteCreate(remoteDir);
    await this.remoteAdd(wsRoot, remoteDir);
  }

  static async createRepoWithReadme(root: string) {
    const git = new Git({ localUrl: root });
    await git.init();
    const readmePath = path.join(root, "README.md");
    fs.ensureFileSync(readmePath);
    await git.add(".");
    await git.commit({ msg: "init" });
  }
}
