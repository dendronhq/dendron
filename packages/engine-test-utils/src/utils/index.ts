import {
  CONSTANTS,
  VaultUtils,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings,
  WorkspaceType,
  normalizeUnixPath,
  ConfigService,
  URI,
} from "@dendronhq/common-all";
import { AssertUtils } from "@dendronhq/common-test-utils";
import { WorkspaceUtils } from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";

export * from "./git";
export * from "./seed";
export * from "./unified";

export async function checkString(body: string, ...match: string[]) {
  return expect(
    await AssertUtils.assertInString({
      body,
      match,
    })
  ).toBeTruthy();
}

export async function checkDir(
  { fpath, snapshot }: { fpath: string; snapshot?: boolean; msg?: string },
  ...match: string[]
) {
  const body = fs.readdirSync(fpath).join(" ");
  if (snapshot) {
    expect(body).toMatchSnapshot();
  }
  return checkString(body, ...match);
}

export async function checkNotInDir(
  { fpath, snapshot }: { fpath: string; snapshot?: boolean; msg?: string },
  ...match: string[]
) {
  const body = fs.readdirSync(fpath).join(" ");
  if (snapshot) {
    expect(body).toMatchSnapshot();
  }
  return checkNotInString(body, ...match);
}

export async function checkFile(
  {
    fpath,
    snapshot,
    nomatch,
  }: { fpath: string; snapshot?: boolean; nomatch?: string[] },
  ...match: string[]
) {
  const body = fs.readFileSync(fpath, { encoding: "utf8" });
  if (snapshot) {
    expect(body).toMatchSnapshot();
  }
  await checkString(body, ...match);
  return !nomatch || checkNotInString(body, ...nomatch);
}

export async function checkNotInString(body: string, ...nomatch: string[]) {
  expect(
    await AssertUtils.assertInString({
      body,
      nomatch,
    })
  ).toBeTruthy();
}

/** The regular version of this only works in engine tests. If the test has to run in the plugin too, use this version. Make sure to check the return value! */
export async function checkFileNoExpect({
  fpath,
  nomatch,
  match,
}: {
  fpath: string;
  nomatch?: string[];
  match?: string[];
}) {
  const body = await fs.readFile(fpath, { encoding: "utf8" });
  return AssertUtils.assertInString({ body, match, nomatch });
}

const getWorkspaceFolders = (wsRoot: string) => {
  const wsPath = path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME);
  const settings = fs.readJSONSync(wsPath) as WorkspaceSettings;
  return _.toArray(settings.folders);
};

export async function checkVaults(opts: WorkspaceOpts, expect: any) {
  const { wsRoot, vaults } = opts;
  const vaultsConfig = (
    await ConfigService.instance().getConfig(
      URI.file(wsRoot),
      "workspace.vaults"
    )
  )._unsafeUnwrap();
  expect(_.sortBy(vaultsConfig, ["fsPath", "workspace"])).toEqual(
    _.sortBy(vaults, ["fsPath", "workspace"])
  );
  if (
    (await WorkspaceUtils.getWorkspaceTypeFromDir(wsRoot)) ===
    WorkspaceType.CODE
  ) {
    const wsFolders = getWorkspaceFolders(wsRoot);
    expect(wsFolders).toEqual(
      vaults.map((ent) => {
        const out: WorkspaceFolderRaw = {
          path: normalizeUnixPath(VaultUtils.getRelPath(ent)),
        };
        if (ent.name) {
          out.name = ent.name;
        }
        return out;
      })
    );
  }
}
