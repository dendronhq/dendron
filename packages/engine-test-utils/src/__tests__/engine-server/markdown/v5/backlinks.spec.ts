import {
  ConfigUtils,
  DVaultVisibility,
  NoteUtils,
  WorkspaceOpts,
} from "@dendronhq/common-all";
import { note2File } from "@dendronhq/common-server";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { DendronASTDest, MDUtilsV5, ProcFlavor } from "@dendronhq/unified";
import { TestConfigUtils } from "../../../../config";
import { ENGINE_HOOKS, ENGINE_HOOKS_MULTI } from "../../../../presets";
import { checkNotInVFile, checkVFile, createProcCompileTests } from "../utils";
import { getOpts, runTestCases } from "./utils";

// NOTE: for setup code details, see, [[../packages/common-test-utils/src/presets/notes.ts#^5xetq2e7t2z4]]
describe("GIVEN dendron.yml default", () => {
  describe("WHEN regular run", () => {
    describe("AND note has backlinks", () => {
      runTestCases(
        createProcCompileTests({
          name: "THEN backlinks are generated",
          fname: "alpha",
          setup: async (opts) => {
            const { proc } = getOpts(opts);
            const resp = await proc.process("");
            return { resp, proc };
          },
          verify: {
            [DendronASTDest.HTML]: {
              [ProcFlavor.PUBLISHING]: async ({ extra }) => {
                const { resp } = extra;
                await checkVFile(resp, "Backlinks", "beta");
              },
            },
          },
          preSetupHook: ENGINE_HOOKS.setupLinks,
        })
      );
    });
  });

  describe("WHEN enableBackLinks = false", () => {
    const setup = async (opts: WorkspaceOpts) => {
      await ENGINE_HOOKS.setupLinks(opts);
      await TestConfigUtils.withConfig(
        (config) => {
          config.publishing!.enableBackLinks = false;
          return config;
        },
        { wsRoot: opts.wsRoot }
      );
    };

    describe("AND WHEN no note override", () => {
      runTestCases(
        createProcCompileTests({
          name: "THEN backlinks are not generated",
          fname: "alpha",
          setup: async (opts) => {
            const { proc } = getOpts(opts);
            const resp = await proc.process("");
            return { resp, proc };
          },
          verify: {
            [DendronASTDest.HTML]: {
              [ProcFlavor.PUBLISHING]: async ({ extra }) => {
                const { resp } = extra;
                await checkNotInVFile(resp, "Backlinks", "beta");
              },
            },
          },
          preSetupHook: async (opts) => {
            return setup(opts);
          },
        })
      );
    });

    describe("GIVEN note override,", () => {
      describe("WHEN note enableBackLinks = true", () => {
        runTestCases(
          createProcCompileTests({
            name: "THEN backlinks are generated",
            fname: "alpha",
            setup: async (opts) => {
              const { proc } = getOpts(opts);
              const resp = await proc.process("");
              return { resp, proc };
            },
            verify: {
              [DendronASTDest.HTML]: {
                [ProcFlavor.PUBLISHING]: async ({ extra }) => {
                  const { resp } = extra;
                  await checkVFile(resp, "Backlinks", "beta");
                },
              },
            },
            preSetupHook: async (opts) => {
              await setup(opts);
              await NoteTestUtilsV4.modifyNoteByPath(
                { ...opts, vault: opts.vaults[0], fname: "alpha" },
                (note) => {
                  NoteUtils.updateNoteLocalConfig(note, "global", {
                    enableBackLinks: true,
                  });
                  return note;
                }
              );
            },
          })
        );
      });

      describe("WHEN note enableBackLinks = false", () => {
        runTestCases(
          createProcCompileTests({
            name: "THEN backlinks are not generated",
            fname: "alpha",
            setup: async (opts) => {
              const { proc } = getOpts(opts);
              const resp = await proc.process("");
              return { resp, proc };
            },
            verify: {
              [DendronASTDest.HTML]: {
                [ProcFlavor.PUBLISHING]: async ({ extra }) => {
                  const { resp } = extra;
                  await checkNotInVFile(resp, "Backlinks", "beta");
                },
              },
            },
            preSetupHook: async (opts) => {
              await setup(opts);
              await NoteTestUtilsV4.modifyNoteByPath(
                { ...opts, vault: opts.vaults[0], fname: "alpha" },
                (note) => {
                  NoteUtils.updateNoteLocalConfig(note, "global", {
                    enableBackLinks: false,
                  });
                  return note;
                }
              );
            },
          })
        );
      });
    });
  });

  describe("WHEN generating backlinks for publishing", () => {
    describe("AND note has backlink to home page", () => {
      runTestCases(
        createProcCompileTests({
          name: "THEN backlinks are generated",
          fname: "beta",
          setup: async (opts) => {
            const { proc } = getOpts(opts);
            const resp = await proc.process("");
            return { resp, proc };
          },
          verify: {
            [DendronASTDest.HTML]: {
              [ProcFlavor.PUBLISHING]: async ({ extra }) => {
                const { resp } = extra;
                await checkVFile(
                  resp,
                  "Backlinks",
                  `<a href="/">Alpha (vault1)</a>`
                );
              },
            },
          },
          preSetupHook: async (opts) => {
            await ENGINE_HOOKS.setupLinks(opts);
            await TestConfigUtils.withConfig(
              (config) => {
                ConfigUtils.setPublishProp(config, "siteHierarchies", [
                  "alpha",
                ]);
                ConfigUtils.setPublishProp(
                  config,
                  "siteUrl",
                  "https://foo.com"
                );
                ConfigUtils.setPublishProp(config, "siteIndex", "alpha");
                return config;
              },
              { wsRoot: opts.wsRoot }
            );
          },
        })
      );
    });
  });

  describe("WHEN generating backlinks to private vaults", () => {
    runTestCases(
      createProcCompileTests({
        name: "THEN private backlinks not added",
        fname: "one",
        setup: async (opts) => {
          const vaults = opts.vaults;
          const { proc } = getOpts(opts);
          const bvault = vaults.find((ent: any) => ent.fsPath === "vault2");
          bvault!.visibility = DVaultVisibility.PRIVATE;

          MDUtilsV5.setProcData(proc, { vaults });
          const resp = await proc.process("");
          return { resp, proc };
        },
        verify: {
          [DendronASTDest.HTML]: {
            [ProcFlavor.PUBLISHING]: async ({ extra }) => {
              const { resp } = extra;
              await checkVFile(
                resp,
                `<a href="/notes/not-secret">Not Secret (vaultThree)</a>`
              );
              await checkNotInVFile(
                resp,
                `<a href="/notes/secret1">Secret1 (vault2)</a>`,
                `<a href="/notes/secret2">Secret2 (vault2)</a>`
              );
            },
          },
        },
        preSetupHook: async (opts: any) => {
          await ENGINE_HOOKS_MULTI.setupBasicMulti(opts);

          await TestConfigUtils.withConfig(
            (config) => {
              const vaults = ConfigUtils.getVaults(config);
              const bvault = vaults.find((ent: any) => ent.fsPath === "vault2");
              bvault!.visibility = DVaultVisibility.PRIVATE;
              const defaultConfig = ConfigUtils.genDefaultConfig();
              ConfigUtils.setVaults(defaultConfig, vaults);
              return defaultConfig;
            },
            { wsRoot: opts.wsRoot }
          );

          const { wsRoot, vaults } = opts;
          await NoteTestUtilsV4.createNote({
            fname: "one",
            vault: vaults[0],
            wsRoot,
            body: "one",
          });
          await NoteTestUtilsV4.createNote({
            fname: "secret1",
            vault: vaults[1],
            wsRoot,
            body: "[[one]]",
          });
          await NoteTestUtilsV4.createNote({
            fname: "secret2",
            vault: vaults[1],
            wsRoot,
            body: "[[one]]",
          });
          await NoteTestUtilsV4.createNote({
            fname: "not-secret",
            vault: vaults[2],
            wsRoot,
            body: "[[one]]",
          });
        },
      })
    );
  });

  describe("WHEN generating backlink to invalid note", () => {
    runTestCases(
      createProcCompileTests({
        name: "THEN do not crash",
        fname: "one",
        setup: async (opts) => {
          const { proc } = getOpts(opts);
          const resp = await proc.process("");
          return { resp, proc };
        },
        verify: {
          [DendronASTDest.HTML]: {
            [ProcFlavor.PUBLISHING]: async ({ extra }) => {
              const { resp } = extra;
              await checkVFile(resp, "Backlinks");
            },
          },
        },
        preSetupHook: async (opts: any) => {
          await ENGINE_HOOKS_MULTI.setupBasicMulti(opts);

          const { wsRoot, vaults } = opts;
          await NoteTestUtilsV4.createNote({
            fname: "one",
            vault: vaults[0],
            wsRoot,
            body: "one",
          });
          await NoteTestUtilsV4.createNote({
            fname: "duplicateOne",
            vault: vaults[1],
            wsRoot,
            body: "[[one]]",
          });

          // Create a note with the same ID as the previous note to create an
          // invalid engine state, and add links to the target note from both
          // notes with the same ID:
          const note = NoteUtils.create({
            created: 1,
            updated: 1,
            id: "duplicateOne",
            fname: "duplicateTwo",
            vault: vaults[1],
            body: "[[one]]",
          });
          await note2File({ note, vault: vaults[1], wsRoot });
        },
      })
    );
  });
});

describe("GIVEN note override,", () => {
  describe("WHEN note enableBackLinks = true", () => {
    runTestCases(
      createProcCompileTests({
        name: "THEN backlinks are generated",
        fname: "alpha",
        setup: async (opts) => {
          const { proc } = getOpts(opts);
          const resp = await proc.process("");
          return { resp, proc };
        },
        verify: {
          [DendronASTDest.HTML]: {
            [ProcFlavor.PUBLISHING]: async ({ extra }) => {
              const { resp } = extra;
              await checkVFile(resp, "Backlinks", "beta");
            },
          },
        },
        preSetupHook: async (opts) => {
          await ENGINE_HOOKS.setupLinks(opts);
          await NoteTestUtilsV4.modifyNoteByPath(
            { ...opts, vault: opts.vaults[0], fname: "alpha" },
            (note) => {
              NoteUtils.updateNoteLocalConfig(note, "global", {
                enableBackLinks: true,
              });
              return note;
            }
          );
        },
      })
    );
  });

  describe("WHEN note enableBackLinks = false", () => {
    runTestCases(
      createProcCompileTests({
        name: "THEN backlinks are not generated",
        fname: "alpha",
        setup: async (opts) => {
          const { proc } = getOpts(opts);
          const resp = await proc.process("");
          return { resp, proc };
        },
        verify: {
          [DendronASTDest.HTML]: {
            [ProcFlavor.PUBLISHING]: async ({ extra }) => {
              const { resp } = extra;
              await checkNotInVFile(resp, "Backlinks", "beta");
            },
          },
        },
        preSetupHook: async (opts) => {
          await ENGINE_HOOKS.setupLinks(opts);
          await NoteTestUtilsV4.modifyNoteByPath(
            { ...opts, vault: opts.vaults[0], fname: "alpha" },
            (note) => {
              NoteUtils.updateNoteLocalConfig(note, "global", {
                enableBackLinks: false,
              });
              return note;
            }
          );
        },
      })
    );
  });
});
