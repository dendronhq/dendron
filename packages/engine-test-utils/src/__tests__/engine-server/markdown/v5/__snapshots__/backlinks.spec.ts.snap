// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN dendron.yml default WHEN enableBackLinks = false AND WHEN no note override "HTML: THEN backlinks are not generated: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"alpha\\">Alpha<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#alpha\\"></a></h1>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`GIVEN dendron.yml default WHEN enableBackLinks = false GIVEN note override, WHEN note enableBackLinks = false "HTML: THEN backlinks are not generated: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"alpha\\">Alpha<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#alpha\\"></a></h1>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`GIVEN dendron.yml default WHEN enableBackLinks = false GIVEN note override, WHEN note enableBackLinks = true "HTML: THEN backlinks are generated: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"alpha\\">Alpha<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#alpha\\"></a></h1>
<hr>
<strong>Backlinks</strong>
<ul>
<li><a href=\\"/notes/beta\\">Beta (vault1)</a></li>
</ul>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`GIVEN dendron.yml default WHEN generating backlink to invalid note "HTML: THEN do not crash: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"one\\">One<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#one\\"></a></h1>
<hr>
<strong>Backlinks</strong>
<ul>
<li><a href=\\"/notes/duplicateOne\\">duplicateTwo (vault2)</a></li>
</ul>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`GIVEN dendron.yml default WHEN generating backlinks for publishing AND note has backlink to home page "HTML: THEN backlinks are generated: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"beta\\">Beta<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#beta\\"></a></h1>
<hr>
<strong>Backlinks</strong>
<ul>
<li><a href=\\"/\\">Alpha (vault1)</a></li>
<li><a title=\\"Private\\" href=\\"https://wiki.dendron.so/notes/hfyvYGJZQiUwQaaxQO27q.html\\" target=\\"_blank\\" class=\\"private\\">Omega (vault1) (Private)</a></li>
</ul>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`GIVEN dendron.yml default WHEN generating backlinks to private vaults "HTML: THEN private backlinks not added: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"one\\">One<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#one\\"></a></h1>
<hr>
<strong>Backlinks</strong>
<ul>
<li><a href=\\"/notes/not-secret.html\\">Not Secret (vaultThree)</a></li>
</ul>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`GIVEN dendron.yml default WHEN generating backlinks to private vaults "HTML: THEN private backlinks not added: PUBLISHING" 2`] = `
VFile {
  "contents": "<h1 id=\\"one\\">One<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#one\\"></a></h1>
<hr>
<strong>Backlinks</strong>
<ul>
<li><a href=\\"/notes/not-secret.html\\">Not Secret (vaultThree)</a></li>
</ul>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`GIVEN dendron.yml default WHEN regular run AND note has backlinks "HTML: THEN backlinks are generated: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"alpha\\">Alpha<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#alpha\\"></a></h1>
<hr>
<strong>Backlinks</strong>
<ul>
<li><a href=\\"/notes/beta\\">Beta (vault1)</a></li>
</ul>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`GIVEN note override, WHEN note enableBackLinks = false "HTML: THEN backlinks are not generated: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"alpha\\">Alpha<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#alpha\\"></a></h1>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`GIVEN note override, WHEN note enableBackLinks = true "HTML: THEN backlinks are generated: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"alpha\\">Alpha<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#alpha\\"></a></h1>
<hr>
<strong>Backlinks</strong>
<ul>
<li><a href=\\"/notes/beta\\">Beta (vault1)</a></li>
</ul>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;
